src/Microsoft.ML.TimeSeries/STL/LeastSquares.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System.Collections.Generic;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.TimeSeries
+{
+    /// <summary>
+    /// this class is used to calculate the least squares of the scatterplots.
+    /// please check http://en.wikipedia.org/wiki/Least_squares for more details.
+    /// </summary>
+    internal class LeastSquares
+    {
+        private readonly IReadOnlyList<double> _x;
+        private readonly IReadOnlyList<double> _y;
+        private readonly int _length;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LeastSquares"/> class.
+        /// constructing the least square algorithm. the input will be consumed directly without any copy, due to memory usage concern.
+        /// </summary>
+        /// <param name="x">the corresponding x-axis value</param>
+        /// <param name="y">the corresponding y-axis value</param>
+        public LeastSquares(IReadOnlyList<double> x, IReadOnlyList<double> y)
+        {
+            Contracts.CheckValue(x, nameof(x));
+            Contracts.CheckValue(y, nameof(y));
+
+            if (x.Count == 0 || y.Count == 0)
+                throw Contracts.Except("input data structure cannot be 0-length");
+            if (x.Count != y.Count)
+                throw Contracts.Except("the x-axis length should be equal to y-axis length!");
+            _x = x;
+            _y = y;
+            _length = _x.Count;
+        }
+
+        /// <summary>
+        /// y=b0+b1x, but the penalty is weighted
+        /// </summary>
+        /// <param name="weights">the weighted least squares. note that the weight should be non-negative, and equal length to data </param>
+        public AbstractPolynomialModel RegressionDegreeOneWeighted(IReadOnlyList<double> weights)
+        {
+            Contracts.CheckValue(weights, nameof(weights));
+
+            Contracts.CheckParam(weights.Count == _length, nameof(weights));
+            if (weights.Count != _length)
+                throw Contracts.Except("the weight vector is not equal length to the data points");
```

Comment:
Suggestion: "The length of the weight vector is not equal to the length of the data points."

Reply:
Updated, thanks.

---
In reply to: [442531436](https://github.com/dotnet/machinelearning/pull/5202#discussion_r442531436) [](ancestors = 442531436)